docs/architecture/adr-008-fee-module.md

```diff
@@ -0,0 +1,95 @@
+# ADR 008: Fee modules
+
+## Changelog
+
+- March 2, 2022: Initial draft.
+
+## Status
+DRAFTED
+
+## Abstract
+This ADR defines the `x/fees` module which enables the possibility to set custom fees for each desmos message
+through governance proposals.
+
+## Context
+In order to better prevent any kind of spam, from zero-gas attack, to smart contracts implementation, to subspaces creation and, 
+in the near future, through posts, it's useful to have a system with which the community can tweak the costs of desmos messages.
+
+## Decision
+We will create a module named `fees` which give desmos community the possibility to add minimum fees to any Desmos custom
+message through governance.
+
+### Types
+Minimum (or custom) fees will contain the following elements:  
+
+* a string identifying the message type delivered with the transaction;
+* the amount of fees associated with the give message.
+
+#### Min Fees
+```go
+type MinFee struct {
+	// The message type identifying the Desmos message
+	messageType string
+	
+	// The amount of fees to be paid for the message
+	amount sdk.Coins
+}
+```
+
+#### Params
+We will save the `MinFee`s for messages in the module's `Params` in order to be able to change them with governance 
+`ParameterChangeProposal`s. To pursue efficiency, the ideal implementation for them is the following:
+
+```go
+type Params struct {
+	MinFees map[string]MinFee
```

Comment:
Considering that
- there won't be too many different messages types supported
- maps are non deterministic inside Go
- the `MinFee` type already contains the message type

We can simply have an array here instead 
```suggestion
	MinFees []MinFee
```

Reply:
I choose this implementation after reading this article:
https://boltandnuts.wordpress.com/2017/11/20/go-slice-vs-maps/
The test-2 shows how it's always better to use map for strings lookup. 
It would make the whole code faster, we will use only one loop I guess and, considering this is executing on the `AnteHandler`, it's better to keep it as fast as possible IMO.

Reply:
@bragaz Ok, that makes sense. But then I would not use `MinFee` as the value type and instead use `sdk.Coins` directly. Otherwise we would have the message URL that is being stored twice each block (as the state is duplicated on each block): both inside the map key and inside the `MinFee` field. So I suggest using something like this 
```go
MinFees map[string]sdk.Coins
```

Where the key represents the Proto URL of the messages, and the `sdk.Coins` represent the min fees for that specific message.

Reply:
@RiccardoM I agree, but I still need to use the `MinFee` type like this:
```protobuf
message MinFee {
  repeated cosmos.base.v1beta1.Coin amount = 2 [
    (gogoproto.nullable) = false,
    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"
  ];
}
```
It's not possible to use repeated types as map's value in protobuf, so I need to wrap it before.

Reply:
Hmm that's a bummer. I really don't like this approach to be honest since it creates overhead just to store something only because we want to use maps rather than lists (there wouldn't be this overhead if we used a lists since the message type would be inside the `MinFee` object itself). 

Let's just step back for a second and consider the usage maps vs lists. 

As you said, this is going to run inside the `AnteHandler`. Currently Desmos supports a blocks size that allows to store up to 50 messages per blocks (this is the maximum amount we've seen during the spam attack). Now, let's consider a 20x worse case: 1.000 messages per block. 

Let's also consider a case in which we set a minimum fee for 100 different types of messages (even though this number will be most likely very smaller). 

Considering [the benchmarks you posted](https://boltandnuts.wordpress.com/2017/11/20/go-slice-vs-maps/), and assuming we are very unlucky and constantly looking for the last element, we would then obtain the following performances: 
- using an array:  
  `128 ns/op * 1.000 = 0.128 ms`
- using a map:  
   `25.7 ns/op * 1.000 = 0.0257 ms`

Even thought the difference is quite significant (using a map allows a 79.9% improvement), the absolute amount of time required to perform such check is extremely small. In both cases we are well under 1 ms. 

Honestly I don't think using a list would cause any problem. On the other hand, using an array would allow us to get the following benefits: 
- a cleaner code base (since we just define a list of objects and put everything inside each object) 
- a more client-friendly setup (since clients would not need to iterate over the key-value pairs to get the data but they would simply have to read the list contents)

Maps are more efficient, I recognize that, but I don't think in our case this would change a lot even in the worst case scenario.

Reply:
Ok that's enough for me! Let's use `MinFee` directly then!